From 225b3a862fb5d12d4f2c42670185cd2856cefd38 Mon Sep 17 00:00:00 2001
From: Frederic Weisbecker <fweisbec@redhat.com>
Date: Tue, 1 Apr 2014 00:04:43 -0400
Subject: [kernel] cputime: Bring cputime -> nsecs conversion

Message-id: <1396310684-6099-3-git-send-email-fweisbec@redhat.com>
Patchwork-id: 78746
O-Subject: [RHEL7 PATCH 2/3] cputime: Bring cputime -> nsecs conversion
Bugzilla: 1047732
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>

Bz: 1047732

Commit d8a9ce3f8ad2b546b9ebaf65de809da0793f11c5
Author: Frederic Weisbecker <fweisbec@gmail.com>
Date: Wed, 5 Mar 2014 16:22:37 +0100

We already have nsecs_to_cputime(). Now we need to be able to convert
the other way around in order to fix a bug on steal time accounting.

Cc: Ingo Molnar <mingo@kernel.org>
Cc: Marcelo Tosatti <mtosatti@redhat.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: Thomas Gleixner <tglx@linutronix.de>
Acked-by: Rik van Riel <riel@redhat.com>
Signed-off-by: Frederic Weisbecker <fweisbec@gmail.com>
Signed-off-by: Frederic Weisbecker <fweisbec@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/asm-generic/cputime_jiffies.h b/include/asm-generic/cputime_jiffies.h
index 272ecba..d5cb78f 100644
--- a/include/asm-generic/cputime_jiffies.h
+++ b/include/asm-generic/cputime_jiffies.h
@@ -15,8 +15,10 @@ typedef u64 __nocast cputime64_t;
 
 
 /*
- * Convert nanoseconds to cputime
+ * Convert nanoseconds <-> cputime
  */
+#define cputime_to_nsecs(__ct)  \
+ jiffies_to_nsecs(cputime_to_jiffies(__ct))
 #define nsecs_to_cputime64(__nsec) \
  jiffies64_to_cputime64(nsecs_to_jiffies64(__nsec))
 #define nsecs_to_cputime(__nsec) \
diff --git a/include/asm-generic/cputime_nsecs.h b/include/asm-generic/cputime_nsecs.h
index 2c9e62c..bf97df3 100644
--- a/include/asm-generic/cputime_nsecs.h
+++ b/include/asm-generic/cputime_nsecs.h
@@ -44,6 +44,8 @@ typedef u64 __nocast cputime64_t;
 /*
  * Convert cputime <-> nanoseconds
  */
+#define cputime_to_nsecs(__ct)  \
+ (__force u64)(__ct)
 #define nsecs_to_cputime(__nsecs) ((__force u64)(__nsecs))
 
 
diff --git a/include/linux/cputime.h b/include/linux/cputime.h
index 2842ebe..f2eb2ee 100644
--- a/include/linux/cputime.h
+++ b/include/linux/cputime.h
@@ -3,6 +3,11 @@
 
 #include <asm/cputime.h>
 
+#ifndef cputime_to_nsecs
+# define cputime_to_nsecs(__ct) \
+ (cputime_to_usecs(__ct) * NSEC_PER_USEC)
+#endif
+
 #ifndef nsecs_to_cputime
 # define nsecs_to_cputime(__nsecs) \
  usecs_to_cputime((__nsecs) / NSEC_PER_USEC)
-- 
1.7.1